File: src/ai/openai.rs
use reqwest;
use serde::{Deserialize, Serialize};
use serde_json;

use crate::ai;
use crate::utils;

#[allow(non_camel_case_types)]
#[derive(Serialize, Deserialize, Debug)]
enum MessageRole {
    system,
    assistant,
    user,
}

#[derive(Serialize, Deserialize, Debug)]
struct Message {
    role: MessageRole,
    content: String,
}

#[derive(Serialize, Deserialize, Debug)]
struct ChatCompletionPayload {
    messages: Vec<Message>,
    model: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    frequency_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    logit_bias: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    logprobs: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    max_tokens: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    n: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    presence_penalty: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    response_format: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    seed: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stop: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tools: Option<Vec<serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    user: Option<String>,
}

impl ChatCompletionPayload {
    fn new(user_message: &str) -> Self {
        let payload = ChatCompletionPayload {
            model: ai::models::GPT4.to_string(),
            messages: vec![
                Message {
                    role: MessageRole::system,
                    content: "You are a helpful chatbot.".to_string(),
                },
                Message {
                    role: MessageRole::user,
                    content: user_message.to_string(),
                },
                Message {
                    role: MessageRole::assistant,
                    content: "The answer (as an integer, and only that integer) is:".to_string(),
                },
            ],
            frequency_penalty: None,
            logit_bias: None,
            logprobs: None,
            max_tokens: Some(40),
            n: Some(1),
            presence_penalty: None,
            response_format: None,
            seed: None,
            stop: None,
            stream: None,
            temperature: Some(1.0),
            top_p: None,
            tools: None,
            user: None,
        };

        payload
    }
}

#[derive(Serialize, Deserialize, Debug)]
struct ChatCompletionResponse {
    choices: Vec<ChatCompletionResponseChoices>,
    created: i64,
    id: String,
    model: String,
    object: String,
    system_fingerprint: String,
    usage: ChatCompletionResponseUsage,
}

#[derive(Serialize, Deserialize, Debug)]
struct ChatCompletionResponseChoices {
    finish_reason: String,
    index: i32,
    logprobs: serde_json::Value,
    message: Message,
}

#[derive(Serialize, Deserialize, Debug)]
struct ChatCompletionResponseUsage {
    completion_tokens: i32,
    prompt_tokens: i32,
    total_tokens: i32,
}

struct Client {
    api_endpoint: String,
    client: reqwest::blocking::Client,
    openai_api_key: String,
}

impl Client {
    fn new() -> Self {
        let client = Client {
            api_endpoint: ai::endpoints::OPENAI_CHAT_COMPLETIONS.to_string(),
            client: reqwest::blocking::Client::new(),
            openai_api_key: utils::env::get_openai_api_key(),
        };

        client
    }

    fn call(&self, payload: &ChatCompletionPayload) -> Option<ChatCompletionResponse> {
        let result = self
            .client
            .post(&self.api_endpoint)
            .header("Content-Type", "application/json")
            .header("Authorization", format!("Bearer {}", self.openai_api_key))
            .json(payload)
            .send();

        match result {
            Ok(response) => {
                let response_obj: ChatCompletionResponse = response.json().unwrap();
                Some(response_obj)
            }
            Err(e) => {
                eprintln!("Request failed: {}", e);
                None
            }
        }
    }
}

pub fn chat_completions(user_message: &str) {
    let client = Client::new();
    let mut payload = ChatCompletionPayload::new(user_message);
    payload.n = Some(3);
    let response = client.call(&payload);

    match response {
        Some(response) => {
            for (i, choice) in response.choices.iter().enumerate() {
                println!("Response {}: {}", i, choice.message.content);
            }

            let pretty = serde_json::to_string_pretty(&response).unwrap();
            println!("{}", pretty);
        }
        None => {
            eprintln!("Failed to get a response from the OpenAI API.");
        }
    }
}

File: src/dkdc/ai/openai.py
# imports
import os
import json
import openai
import tiktoken


from enum import Enum
from pydantic import BaseModel, Field

from dkdc.ai.systems import DKDC_AI, DKDC_CLASSIFY, DKDC_CAST
from dkdc.ai.tokenize import str_to_tokens, get_tokenizer
from dkdc.utils import filesystem as fs
from dkdc.defaults import OPENAI_MODEL
from dkdc.utils.vars import load_vars

OPENAI_MODEL = "llama3.1:70b"


class MessageRole(str, Enum):
    SYSTEM = "system"
    ASSISTANT = "assistant"
    USER = "user"


class Message(BaseModel):
    role: MessageRole
    content: str


class ChatCompletionPayload(BaseModel):
    messages: list[Message]
    model: str
    frequency_penalty: float = None
    logit_bias: dict[int, float] = None
    logprobs: bool = None
    max_tokens: int = None
    n: int = None
    presence_penalty: float = None
    response_format: str = None
    seed: int = None
    stop: str = None
    stream: bool = None
    temperature: float = None
    top_p: float = None
    tools: list[dict] | None = None
    tool_choice: dict | None = None
    user: str = None


class ResponseMessage(BaseModel):
    role: MessageRole
    content: str | None = None
    function_call: str | None = None
    tool_calls: list[dict] | None = None


class ChatCompletionResponseChoices(BaseModel):
    finish_reason: str
    index: int
    logprobs: dict | None = None
    message: ResponseMessage


class ChatCompletionResponseUsage(BaseModel):
    completion_tokens: int
    prompt_tokens: int
    total_tokens: int


class ChatCompletionResponse(BaseModel):
    id: str
    choices: list[ChatCompletionResponseChoices]
    created: int
    model: str
    object: str
    system_fingerprint: str
    usage: ChatCompletionResponseUsage


class Client:
    def __init__(self):
        load_vars()
        sys_msg = Message(role=MessageRole.SYSTEM, content=DKDC_AI)

        # self.client = openai.OpenAI(base_url="")  # api_key=os.getenv("OPENAI_API_KEY"))
        self.client = openai.OpenAI(
            base_url="http://localhost:11434/v1", api_key="ollama"
        )
        self.messages = [sys_msg]
        self.context = {}

    def __call__(self, text: str) -> ChatCompletionResponse:
        usr_msg = Message(role=MessageRole.USER, content=text)

        self.messages.append(usr_msg)

        payload = ChatCompletionPayload(
            messages=self.messages, model=OPENAI_MODEL, temperature=0.0
        )

        response = self.client.chat.completions.create(**payload.dict())
        response = ChatCompletionResponse(**response.dict())

        resp_msg = response.choices[0].message

        self.messages.append(
            Message(
                role=MessageRole.ASSISTANT,
                content=resp_msg.content,
            )
        )

        return response

    def add_context(self, files: dict[str, str]) -> None:
        self.context = self.context | files
        self.messages[0].content += f"\n{fs.files_to_markdown(files)}"

    def reset(self) -> None:
        self.context = {}
        sys_msg = Message(role=MessageRole.SYSTEM, content=DKDC_AI)
        self.messages = [sys_msg]

    def classify(
        self,
        text: str,
        labels: list[str],
        allow_none: bool = False,
        verbose: bool = False,
        n: int = 1,
    ) -> str:
        enc = get_tokenizer()

        if allow_none:
            labels.append("None of the above")

        logit_bias = {}
        for i, _ in enumerate(labels):
            logit_bias[enc.encode(str(i))[0]] = 100.0

        text = f"{text}\nChoices:\n  {labels}"

        if verbose:
            print(f"Text: {text}")

        system = DKDC_CLASSIFY

        for i, label in enumerate(labels):
            system += f"{i}: {label}\n"

        if verbose:
            print(f"System: {system}")

        sys_msg = Message(role=MessageRole.SYSTEM, content=system)
        cls_msg = Message(role=MessageRole.USER, content=text)

        payload = ChatCompletionPayload(
            messages=[sys_msg, cls_msg],
            model=OPENAI_MODEL,
            logit_bias=logit_bias,
            max_tokens=1,
            n=n,
        )

        response = self.client.chat.completions.create(**payload.dict())
        response = ChatCompletionResponse(**response.dict())

        votes = {}
        for i, choice in enumerate(response.choices):
            if verbose:
                print(f"Response {i}:\n{choice.message.content}")
                print(f"Chose category: {labels[int(choice.message.content)]}")

            category = int(choice.message.content)
            count = votes.get(category, 0)
            votes[category] = count + 1

        if verbose:
            print(response.json())

        max_label = max(votes, key=votes.get)
        return labels[max_label]

    def cast(
        self,
        text: str,
        model_class: BaseModel,
        additional_instructions: str = None,
    ):
        system = DKDC_CAST
        if additional_instructions:
            system += f"### additional instructions\n\n{additional_instructions}"

        sys_msg = Message(role=MessageRole.SYSTEM, content=system)
        cls_msg = Message(role=MessageRole.USER, content=text)

        tools = [
            {
                "type": "function",
                "function": {
                    "name": "cast",
                    "description": "cast text into a pydantic model",
                    "parameters": model_class.schema(),
                },
            }
        ]
        tool_choice = {"type": "function", "function": {"name": "cast"}}

        payload = ChatCompletionPayload(
            messages=[sys_msg, cls_msg],
            model=OPENAI_MODEL,
            tools=tools,
            tool_choice=tool_choice,
            temperature=0.0,
        )

        response = self.client.chat.completions.create(**payload.dict())
        response = ChatCompletionResponse(**response.dict())

        class_output = json.loads(
            response.choices[0].message.tool_calls[0]["function"]["arguments"]
        )

        return model_class(**class_output)
